### PR TITLE
rollup: add JSDoc, spacesToTabs fix, improved debugging builds

### PR DESCRIPTION
**Pull request by kungfooman**
I would like to suggest some changes I figured over time:

1) `spacesToTabs(enable)` requires `true` as argument, or it simply does nothing
2) `debugger` statements were stripped by the strip plugin, even in debug mode, which annoys and surprises me every time again. Since no function is stripped in Debug mode, I just enable the `strip` plugin for the other modes.
3) The regex in `spacesToTabs` never possibly could work, hence no output file actually contained tabs. Even if regex would work, it would still be a dangerous regex, that doesn't respect "line start spaces" only, so it would e.g. even replace spaces inside a string.
4) I added JSDoc with IntelliSense (the `jsconfig.json` changes are required for this - as far I tested it doesn't affect anything else  so far)
5) Plugins needs to specify a `name`, it is a mandatory property, otherwise Babel error messages cannot point out the name of the plugin that encountered an exception.
6) I would like to keep comments in my debug builds, they always help me to understand issues while debugging

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by willeastcott** [comment]
This looks good. But can you please ensure there's a newline before every JSDoc block please?

For comments in the debug build, didn't we make it so source maps are now available with engine builds now?

File: rollup.config.mjs [hunk 1]
```diff
@@ -14,12 +14,24 @@ import dts from 'rollup-plugin-dts';
 import jscc from 'rollup-plugin-jscc';
 import { visualizer } from 'rollup-plugin-visualizer';
 
+/** @typedef {import('rollup'              ).RollupOptions                } RollupOptions                 */
+/** @typedef {import('rollup'              ).Plugin                       } Plugin                        */
+/** @typedef {import('rollup'              ).OutputOptions                } OutputOptions                 */
+/** @typedef {import('@rollup/plugin-babel').RollupBabelInputPluginOptions} RollupBabelInputPluginOptions */
+/** @typedef {import('@rollup/plugin-strip').RollupStripOptions           } RollupStripOptions            */
```
Comment (willeastcott) [inline rollup.config.mjs:21]: /** @typedef {import('rollup').RollupOptions} RollupOptions */
/** @typedef {import('rollup').Plugin} Plugin */
/** @typedef {import('rollup').OutputOptions} OutputOptions */
/** @typedef {import('@rollup/plugin-babel').RollupBabelInputPluginOptions} RollupBabelInputPluginOptions */
/** @typedef {import('@rollup/plugin-strip').RollupStripOptions} RollupStripOptions */

Comment (willeastcott) [reply rollup.config.mjs:21]: I know this is visually pleasing but we don't align imports/typedefs/etc like there elsewhere in the codebase, so for the purpose of standardization, I've suggested the above.

File: jsconfig.json [hunk 2]
```diff
@@ -3,7 +3,8 @@
...
         "checkJs": true,
         "module": "es6",
-        "target": "es6"
+        "target": "es6",
+        "moduleResolution": "node"
     },
-    "include": ["src/**/*"]
-}
\ No newline at end of file
+    "include": ["src/**/*", "rollup.config.js"],
```
Comment (willeastcott) [inline jsconfig.json:9]:
```suggestion
    "include": ["src/**/*", "rollup.config.mjs"]
```

**Comment by kungfooman** [comment]
> For comments in the debug build, didn't we make it so source maps are now available with engine builds now?

I just checked if sourcemaps still work and they cause no problems with comments:

image

Basically just `npm run build:sourcemaps` and then clicking on `pc.Entity.prototype.translate` in devtools.

Were you wondering about this, if it still works?

**Comment by willeastcott** [comment]
No, what I meant was, why keep the comments in the debug build when you can just use source maps and see the original comments?

**Comment by kungfooman** [comment]
> No, what I meant was, why keep the comments in the debug build when you can just use source maps and see the original comments?

Oh, that's what you mean, I don't really like sourcemaps for debugging. If you change/fix some code directly in the build then you won't even see your changes through the "sourcemap lense".

So what I can do right now is to just make a small fix to `rollup.config.mjs` to ignore inline debug sourcemaps and then I am free to edit the debug output files again, without sourcemaps messing around with my changes.

I realize my work flow is maybe a bit strange, but I just try to not rebuild everything all the time, since it takes quite some time.

Then the argument is, that I could also use the watch-mode, but even though it's quicker, it still takes like three seconds - for updating one file change. And then always these background tasks you need to deal with (which block my default VSCode shell aswell... ending up creating multiple ones that I need to cognize), I feel way too impatient for all that.

**Review by mvaligursky — APPROVED** [review]
This all looks great to me! 
Any further thoughts on this @willeastcott ?

**Review by willeastcott — APPROVED** [review]

**Comment by willeastcott** [comment]
@mvaligursky LGTM.

**Comment by mvaligursky** [comment]
Thanks for an awesome contribution @kungfooman !

**Pull request merged**